### PR TITLE
Secret volume plugin iteration 1

### PR DESCRIPTION
**Pull request by pmorie**
Split off from the prototype developed with #4126, contains initial implementation of secret volume plugin.  Next steps: tmpfs filesystem for secrets on node.

### CODE REVIEW

**Comment by pmorie** [comment]
@erictune Ready for a review of the generalities / e2e test, but still in need of tmpfs-related changes

**Comment by erictune** [comment]
@thockin or @dchen1107 might have a look at this.

**Comment by erictune** [comment]
@pmorie rebase now that other two PRs merged.

**Comment by erictune** [comment]
Some comments in https://github.com/pmorie/kubernetes/commit/c80ec0a27bfd1d7e8c6d898cf668af9dcb543959

**Comment by pmorie** [comment]
@erictune rebased

File: pkg/kubelet/volume/empty_dir/empty_dir_test.go [hunk 1]
```diff
@@ -27,7 +27,7 @@ import (
 
 func TestCanSupport(t *testing.T) {
 	plugMgr := volume.PluginMgr{}
-	plugMgr.InitPlugins(ProbeVolumePlugins(), &volume.FakeHost{"/tmp/fake"})
+	plugMgr.InitPlugins(ProbeVolumePlugins(), &volume.FakeHost{"/tmp/fake", nil})
```
Comment (erictune) [inline pkg/kubelet/volume/empty_dir/empty_dir_test.go]: Are these changes actually necessary? Doesn't go initialize an unspecified value to nil?

File: pkg/kubelet/volume/secret/secret.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+func (plugin *secretPlugin) newCleanerInternal(volName string, podUID types.UID) (volume.Cleaner, error) {
+	return &secretVolume{volName, podUID, plugin, nil}, nil
+}
+
+// secretVolume handles retrieving secrets from the API server
+// and placing them into the volume on the host.
+type secretVolume struct {
+	volName   string
+	podUID    types.UID
+	plugin    *secretPlugin
```
Comment (erictune) [inline pkg/kubelet/volume/secret/secret.go]: Instead of storing a pointer to the plugin in the secretVolume, suggest just storing it in a string in secretVolume and computing it when you initialize secretVolume{} in newCleaner and newBuilder.

This makes it so that secretVolume only has the information it needs. Also, its not immediately obvious that GetPath() would always return the same value, whereas we don't expect the value to change over the lifetime of the Builder/Cleaner objects.

File: pkg/kubelet/volume/empty_dir/empty_dir_test.go (hunk 1 above)
Comment (pmorie) [inline pkg/kubelet/volume/empty_dir/empty_dir_test.go]: @erictune you're right, I'm not sure why I did that.

Comment (pmorie) [inline pkg/kubelet/volume/empty_dir/empty_dir_test.go]: @erictune I remember why now; if you use field initializers you have to initialize all the fields.  If you just do:

```go
volume.FakeHost{"/tmp/fake"}
```

you get a too-few fields in initializer error.  If you don't want to supply the `nil`, you have to do:

```go
volume.FakeHost{RootDir: "/tmp/fake"}
```

I'm open to changing it or making a factory method.

Comment (erictune) [inline pkg/kubelet/volume/empty_dir/empty_dir_test.go]: no, leave it as is.

File: pkg/kubelet/volume/secret/secret.go (hunk 2 above)
Comment (pmorie) [inline pkg/kubelet/volume/secret/secret.go]: The pointer is stored because the `Setup` method uses the KubeClient it provides to fetch the secrets from the API server.

**Comment by pmorie** [comment]
@erictune Do you want the tmpfs changes in this PR or in another one?

**Comment by erictune** [comment]
no preference.

**Comment by pmorie** [comment]
@erictune I don't want the keys in a secret to be coupled to being paths.  I'd prefer to use an existing format, but I don't think C_IDENTIFIER or DNS_label is quite appropriate.  I want users to be able to use `-` and `_` and to have filename like constructions (viz: `id_rsa.pub`), but definitely not path separators.  Thoughts on the following format?

    [a-zA-Z][a-zA-Z0-9_-]*\.?[a-zA-Z0-9]

**Comment by pmorie** [comment]
> no preference.

I think I would prefer to get the generalities in first in this PR and do tmpfs as a follow-up.

**Comment by pmorie** [comment]
@erictune something like v1beta3 DNS_SUBDOMAIN but with `_`

**Comment by pmorie** [comment]
@erictune I'm inclined for now to say that keys have to be a DNS_SUBDOMAIN, and we can make the SecretSource and volume plugin adapt things into filenames if people need fine grained control over file names.

**Comment by pmorie** [comment]
@erictune This is in the state I want it for this PR.  Will do tmpfs and fine-grained control of paths for files in volumes in subsequent PRs.

**Comment by pmorie** [comment]
@erictune My e2e test passes locally.  I'm doing a run in GCE now.

**Comment by pmorie** [comment]
@erictune My e2e test passes in GCE as well.

File: pkg/api/types.go [hunk 3]
```diff
@@ -230,7 +230,10 @@ type GitRepo struct {
 	// TODO: Consider credentials here.
 }
 
-// Adapts a Secret into a VolumeSource
+// Adapts a Secret into a VolumeSource.
+//
+// The contents of the target Secret's Data field will be presented in a volume
+// as files using the keys in the Data field as the file names.
 type SecretSource struct {
```
Comment (thockin) [inline pkg/api/types.go]: Did we consider SecretSource holding a slice of ObjectReferences so you could pack more than one secret into a single mount?
